### PR TITLE
store/retrieve lifecycle states in user operations

### PR DESCRIPTION
**Pull request by IsuraD**

### CODE REVIEW

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/executor/UserLifeCycleExecutor.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/***********************************************************************************************************************
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/executor/UserLifeCycleExecutor.java]: Wrong header

File: feature/org.wso2.carbon.identity.mgt.feature/pom.xml [hunk 2]
```diff
@@ -74,6 +74,7 @@
                                         <include>dbscripts/identity-mgt/mysql.sql</include>
                                         <include>p2.inf</include>
                                         <include>build.properties</include>
+                                        <include>conf/UserLifeCycle.xml</include>
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/pom.xml]: Shall we use a lowercase file name like `user-life-cycle.xml`. All the other confs are in this format.

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/executor/UserLifeCycleExecutor.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * @param resource     The resource in which the lifecycle state is changed.
+     * @param currentState Current lifecycle state.
+     * @param targetState  The target lifecycle state.
+     * @throws LifecycleException If exception occurs while running the executor.
+     */
+    @Override
+    public void execute(Object resource, String currentState, String targetState) throws LifecycleException {
+        //
+        log.info(currentState);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/executor/UserLifeCycleExecutor.java]: Is this needed ?

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java [hunk 4]
```diff
@@ -37,4 +37,5 @@
     public static final String USERSTORE_DISPLAY_NAME = "display-name";
     public static final String USER_EXIST = "userExist";
     public static final String NO_OF_DOMAINS = "noOfDomains";
+    public static final String USER_LIFE_CYCLE = "UserLifeCycle";
```
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java]: Let's change the value to `userLifeCycle` to maintain a uniformity.

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                    </datamodel>
+                    <transition event="Locked invalid credentials -verified" target="LOCKED_INVALID_CREDENTIAL#UNVERIFIED"/>
+                    <transition event="Locked invalid asnwer -verified" target="LOCKED_INVALID_ANSWER#UNVERIFIED"/>
+                    <transition event="Disabled" target="DISABLED"/>
+                </state>
+
+                <state id="UNLOCKED#VERIFIED">
+                    <datamodel>
+                        <data name="transitionExecution">
+                            <execution forTarget="LOCKED_INVALID_CREDENTIAL#VERFIED"
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: typo ? should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                    <datamodel>
+                        <data name="transitionExecution">
+                            <execution forTarget="LOCKED_INVALID_CREDENTIAL#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_INVALID_ANSWER#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+			    <execution forTarget="DISABLED"
+                                       class="org.wso2.carbon.identity.handler.event.acco	unt.lock.executors.APIExecutor">
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: `acco	unt` should be `account` ?

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java (hunk 4 above)
Comment (IsuraD) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java]: I don't think we need to change this. It is comming from the XML

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                            </execution>
+                        </data>
+                    </datamodel>
+                    <transition event="Locked invalid credentials -verified" target="LOCKED_INVALID_CREDENTIAL#UNVERIFIED"/>
+                    <transition event="Locked invalid asnwer -verified" target="LOCKED_INVALID_ANSWER#UNVERIFIED"/>
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: typo ? should be `answer` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+                <state id="UNLOCKED#VERIFIED">
+                    <datamodel>
+                        <data name="transitionExecution">
+                            <execution forTarget="LOCKED_INVALID_CREDENTIAL#VERFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_INVALID_ANSWER#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_AIPR#VERFIED"
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+			    <execution forTarget="DISABLED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                        </data>
+                    </datamodel>
+                    <transition event="Locked invalid credentails -verified" target="LOCKED_INVALID_CREDENTIAL#VERFIED"/>
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+   		<state id="DISABLED">
+                    <datamodel>
+                        <data name="transitionExecution">
+                            <execution forTarget="UNLOCKED#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="UNLOCKED#VERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_AIPR#VERFIED"
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_AIPR#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                            <execution forTarget="LOCKED_INVALID_CREDENTIAL#VERFIED"
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+			    <execution forTarget="LOCKED_INVALID_ANSWER#UNVERIFIED"
+                                       class="org.wso2.carbon.identity.mgt.executor.UserLifeCycleExecutor">
+                            </execution>
+                        </data>
+                    </datamodel>
+                    <transition event="Unlocked -unverified" target="UNLOCKED#UNVERIFIED"/>
+                    <transition event="Unlocked -verified" target="UNLOCKED#VERIFIED"/>
+                    <transition event="Locked -verified" target="LOCKED_AIPR#VERFIED"/>
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+                    <transition event="Locked -unverified" target="LOCKED_AIPR#UNVERIFIED"/>
+                    <transition event="Locked invalid credentails -verified" target="LOCKED_INVALID_CREDENTIAL#VERFIED"/>
```
Comment (omindu) [inline feature/org.wso2.carbon.identity.mgt.feature/resources/conf/UserLifeCycle.xml]: should be `VERIFIED` ?

File: components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java (hunk 4 above)
Comment (omindu) [inline components/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/constant/IdentityMgtConstants.java]: Okay.

**Pull request merged**